Applications/Spire/Source/Ui/ColorSelectorDropDown.cpp

```diff
@@ -0,0 +1,243 @@
+    m_color_hue_slider->set_color(color);
+    on_color_selected(color);
+  });
+  color_picker_layout->addWidget(m_hex_input);
+  color_picker_layout->addStretch(1);
+  auto recent_colors_label = new QLabel(tr("Recent Colors"), this);
+  recent_colors_label->setFixedHeight(scale_height(LABEL_HEIGHT()));
+  recent_colors_label->setStyleSheet(label_style);
+  layout->addWidget(recent_colors_label);
+  m_recent_colors_layout = new QHBoxLayout();
+  m_recent_colors_layout->setContentsMargins(0, scale_height(10), 0, 0);
+  m_recent_colors_layout->setSpacing(HORIZONTAL_PADDING());
+  layout->addLayout(m_recent_colors_layout);
+  update_recent_colors_layout();
+  m_recent_colors.connect_change_signal([=] { on_recent_colors_changed(); });
+}
+
+void ColorSelectorDropDown::set_color(const QColor& color) {
+  m_color_hue_slider->set_color(color);
+  m_color_value_slider->set_color(color);
+  m_hex_input->set_color(color);
+}
+
+connection ColorSelectorDropDown::connect_color_signal(
+    const ColorSignal::slot_type& slot) const {
+  return m_color_signal.connect(slot);
+}
+
+void ColorSelectorDropDown::childEvent(QChildEvent* event) {
+  if(event->added()) {
+    event->child()->installEventFilter(this);
+  }
+}
+bool ColorSelectorDropDown::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::KeyPress) {
+    auto e = static_cast<QKeyEvent*>(event);
+    if(e->key() == Qt::Key_Escape) {
+      hide();
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void ColorSelectorDropDown::hideEvent(QHideEvent* event) {
+  if(m_current_color != m_original_color) {
+    m_recent_colors.add_color(m_current_color);
+    m_original_color = m_current_color;
+    update_recent_colors_layout();
+  }
+}
+
+void ColorSelectorDropDown::add_basic_color_button(int x, int y,
+    const QColor& color) {
+  auto button = create_color_button(color, this);
+  button->connect_clicked_signal([=, color = std::move(color)] {
+    on_color_button_clicked(color);
+  });
+  m_basic_colors_layout->addWidget(button, x, y);
+}
+
+void ColorSelectorDropDown::add_recent_color_button(int index,
+    const QColor& color) {
+  auto button = create_color_button(color, this);
+  button->connect_clicked_signal([=, color = std::move(color)] {
+    on_color_button_clicked(color);
+  });
+  m_recent_colors_layout->insertWidget(index, button);
+}
+
+void ColorSelectorDropDown::update_recent_colors_layout() {
+  while(m_recent_colors_layout->itemAt(0) != nullptr) {
+    auto item = m_recent_colors_layout->takeAt(0);
+    delete item->widget();
+    delete item;
+  }
+  auto colors = m_recent_colors.get_colors();
+  for(auto i = std::size_t(0); i < colors.size(); ++i) {
+    add_recent_color_button(i, colors[i]);
+  }
+}
+
+void ColorSelectorDropDown::on_color_button_clicked(const QColor& color) {
+  set_color(color);
+  on_color_selected(color);
+}
+
+void ColorSelectorDropDown::on_color_selected(const QColor& color) {
+  m_color_signal(color);
```

Comment:
This violates principles of re-entrancy. You must only signal to the outside world that a change has been made after you make the change, not before. Anyone listening to the color signal will see inconsistencies between what color was selected and the current color.